csharp/ql/test/library-tests/dataflow/tuples/Tuples.cs

```diff
@@ -0,0 +1,84 @@
+using System;
+
+class Tuples
+{
+    static void M1()
+    {
+        var x = (a: "taint source", (1, "taint source"));
+        var (a, (b, c)) = x;
+        Sink(a);                // Tainted
+        Sink(b);
+        Sink(c);                // Tainted
+
+        (a, (b, c)) = x;
+        Sink(a);                // Tainted
+        Sink(b);
+        Sink(c);                // Tainted
+
+        (var p, var q) = x;
+        Sink(p);                // Tainted
+        Sink(q.Item1);
+        Sink(q.Item2);          // Tainted
+
+        Sink(x.Item1);          // Tainted
+        Sink(x.a);              // Tainted
+        Sink(x.Item2.Item1);
+        Sink(x.Item2.Item2);    // Tainted
+    }
+
+    static void M2()
+    {
+        var x = ("taint source", 2, 3, 4, 5, 6, 7, 8, 9, "taint source");
+        Sink(x.Item1);          // Tainted
+        Sink(x.Item2);
+        Sink(x.Item10);         // Tainted
+    }
+
+    static void M3()
+    {
+        var x = (ValueTuple<string, int, int>)("taint source", 2, 3);
+        Sink(x.Item1);          // Tainted
```

Comment:
I'm a bit surprised that this is found.

Reply:
Casts are modelled as normal value-preserving data-flow steps: https://github.com/github/codeql/blob/main/csharp/ql/src/semmle/code/csharp/dataflow/internal/DataFlowPrivate.qll#L188.